dlme_airflow/drivers/json.py

```diff
@@ -1,155 +1,246 @@
+                    ][0]
+                    while next:
+                        self._page_urls.append(next)
+
+                        resp = self._get(next)
+                        if resp.status_code == 200:
+                            result = resp.json()
+                            next = [match.value for match in expression.find(result)][0]
+                        else:
+                            logging.error(
+                                f"got {resp.status_code} when fetching {next}"
+                            )
+                else:
+                    raise Exception(
+                        f"Pagination path not found in: {self.collection_url}"
+                    )
+            else:
+                logging.error(
+                    f"got {resp.status_code} when fetching {self.collection_url}"
+                )
+        else:
+            self._page_urls.append(self.collection_url)
+            logging.info(f"getting collection {self.collection_url}")
+            # pass a json path for the result count in the catalog or hard code the count
+            if self.result_count:
+                if self.result_count.get("path"):
+                    resp = self._get(self.collection_url)
+                    if resp.status_code == 200:
+                        collection_result = resp.json()
+                        expression = jsonpath_ng.parse(self.result_count.get("path"))
+
+                        if [
+                            match.value for match in expression.find(collection_result)
+                        ]:
+                            result_count = [
+                                match.value
+                                for match in expression.find(collection_result)
+                            ][0]
+                else:
+                    try:
+                        result_count = int(self.result_count.get("count"))
+                    except result_count.does_not_exist:
+                        logging.error(
+                            "result_count needs to be set in the catalog with a json path or a hard coded number"
+                        )
+            url = self.collection_url
+            if self.increment:
+                start = int(self.increment.get("amount"))
+                for i in range(
+                    int(self.increment.get("amount")),
+                    result_count,
+                    int(self.increment.get("amount")),
+                ):
+                    url = url.replace(
+                        f"{self.increment.get('variable_string')}{start}",
+                        f"{self.increment.get('variable_string')}{start+int(self.increment.get('amount'))}",
+                    )
+                    start += int(self.increment.get("amount"))
+                    self._page_urls.append(url)
+            else:
+                logging.error("increment key not found in catalog")
+
+    def _open_page(self, page_url: str) -> Optional[list]:
```

Comment:
I'm not sure if this is worth the complexity. Its nice to have one driver for json collections but to do that, the driver has to handle different kinds of paging options. This works for LoC and Met. We have 2 other json collections. One won't work with this driver because you have to hit the API to get a set of ids first and then harvest the metadata. The other one seems to have changed our permissions so we can't access the API currently.

Reply:
I think the first step will be to DRY up this section through some refactoring. I think that will determine _if_ we can have a single driver. Taking a pass at that now.